### PR TITLE
bug 7242 app-layer proto negation

### PR DESCRIPTION
**Pull request by victorjulien**
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2043

https://redmine.openinfosecfoundation.org/issues/7242

### CODE REVIEW

File: src/detect-app-layer-protocol.c [hunk 1]
```diff
@@ -214,15 +215,15 @@ PrefilterPacketAppProtoMatch(DetectEngineThreadCtx *det_ctx, Packet *p, const vo
...
+        int r = AppProtoEquals(ctx->v1.u16[0], p->flow->alproto_ts) ^ ctx->v1.u8[2];
         if (r) {
             PrefilterAddSids(&det_ctx->pmq, ctx->sigs_array, ctx->sigs_cnt);
         }
-    } else if ((p->flags & PKT_PROTO_DETECT_TC_DONE) && (p->flowflags & FLOW_PKT_TOCLIENT))
-    {
-        int r = (ctx->v1.u16[0] == p->flow->alproto_tc) ^ ctx->v1.u8[2];
+    } else if ((p->flags & PKT_PROTO_DETECT_TC_DONE) && (p->flowflags & FLOW_PKT_TOCLIENT) &&
+               p->flow->alproto_tc != ALPROTO_UNKNOWN) {
+        int r = AppProtoEquals(ctx->v1.u16[0], p->flow->alproto_tc) ^ ctx->v1.u8[2];
```
Comment (catenacyber) [inline src/detect-app-layer-protocol.c:226]: nit: `r` could be a bool by the way

File: src/detect-app-layer-protocol.c [hunk 2]
```diff
@@ -214,15 +215,15 @@ PrefilterPacketAppProtoMatch(DetectEngineThreadCtx *det_ctx, Packet *p, const vo
         SCReturn;
     }
 
-    if ((p->flags & PKT_PROTO_DETECT_TS_DONE) && (p->flowflags & FLOW_PKT_TOSERVER))
-    {
-        int r = (ctx->v1.u16[0] == p->flow->alproto_ts) ^ ctx->v1.u8[2];
+    if ((p->flags & PKT_PROTO_DETECT_TS_DONE) && (p->flowflags & FLOW_PKT_TOSERVER) &&
+            p->flow->alproto_ts != ALPROTO_UNKNOWN) {
+        int r = AppProtoEquals(ctx->v1.u16[0], p->flow->alproto_ts) ^ ctx->v1.u8[2];
```
Comment (catenacyber) [inline src/detect-app-layer-protocol.c:220]: Code looks good to me.

How do you test this `AppProtoEquals` change ? 
Use of prefilter and `http` without version specified ?

Comment (victorjulien) [reply src/detect-app-layer-protocol.c:220]: I didn't test it :)

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
  

Pipeline 22541

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 22596

**Review by jasonish — APPROVED** [review]
Looks OK. And no commit ID's as it wasn't a cherry pick.

**Comment by victorjulien** [comment]
Merged in #11789, thanks!

**Pull request merged**